modules/agar/src/test/ts/browser/ApproxStructureTest.ts

```diff
@@ -1,139 +1,426 @@
+            s.elementWithSpecifics('div', {}),
+            s.zeroOrOne(s.elementWithSpecifics('span', {
+              classes: [
+                'bye'
+              ]
+            }))
+          ]
+        }));
+
+      check(struct2, '<div><span class="hello"></span><div></div></div>');
+      check(struct2, '<div><span class="hello"></span><div></div><span class="bye"></span></div>');
+      check(struct2, '<div><span class="hello"></span><span class="hello"></span><span class="hello"></span><span class="hello"></span><div></div><span class="bye"></span></div>');
+    });
+
+    it('TINY-9102: ApproxStructure text combineSiblings', () => {
+      const container = SugarElement.fromTag('div');
+      InsertAll.append(container, [
+        SugarElement.fromText('hello'),
+        SugarElement.fromText(' '),
+        SugarElement.fromText('world')
+      ]);
+
+      Assertions.assertStructure('Test', ApproxStructure.build((s, str, _arr) =>
+        s.elementWithSpecifics('div', {
+          children: [
+            s.text(str.is('hello world'), true)
+          ]
+        })), container);
+
+      Assertions.assertStructure('Test', ApproxStructure.build((s, str, _arr) =>
+        s.elementWithSpecifics('div', {
+          children: [
+            s.text(str.is('hello'), false),
+            s.text(str.is(' world'), true)
           ]
-        })),
-        s.element('div', {}),
-        s.zeroOrOne(s.element('span', {
+        })), container);
+    });
+
+    it('TINY-9102: ApproxStructure throws error when there are extra attributes', () => {
+      const html = '<div extra-attribute2="extra2" extra-attribute="extra" data-key="test-1" selected="double" class="test1 root" style="display: block;">' +
+        '</div>';
+
+      checkThrowError(ApproxStructure.build((s, str, _arr) =>
+        s.elementWithSpecifics('div', {
+          attrs: {
+            'selected': str.is('double'),
+            'data-key': str.contains('test'),
+            'extra-attribute': str.startsWith('ex'),
+            'non-existent-attribute': str.is('non'),
+          },
+          classes: [
+            'test1', 'root'
+          ],
+          styles: {
+            display: str.is('block')
+          },
+        })), html);
+    });
+
+    it('TINY-9102: ApproxStructure throws error when there are extra classes', () => {
+      const html = '<div data-key="test-1" selected="double" class="extra-class test1 root" style="display: block;">' +
+        '</div>';
+
+      checkThrowError(ApproxStructure.build((s, str, _arr) =>
+        s.elementWithSpecifics('div', {
+          attrs: {
+            'selected': str.is('double'),
+            'data-key': str.contains('test')
+          },
+          classes: [
+            'test1', 'root', 'non-existent-class'
+          ],
+          styles: {
+            display: str.is('block')
+          },
+        })), html);
+    });
+
+    it('TINY-9102: ApproxStructure throws error when there are extra styles', () => {
```

Comment:
It looks like we don't have any tests to check that if there aren't enough classes/styles/attributes, it still fails. Is that correct?

Reply:
Thanks, yeah, I have just added it after you commented. It should also throw errors when there are fewer classes/styles/attributes.